dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Routes all requests to one host until an error is received, then we'll move on to the next host.
+ * Upside: ensures clients will always hit warm caches
+ * Downsides:
+ * <ul>
+ *     <li>a single high volume client can result in unbalanced utilization of server nodes.
+ *     <li>when a node is restarted (and clients see 500s/connect exceptions), they will _all_ pin to a different node
+ * </ul>
+ *
+ * To alleviate the second downside, we reshuffle all nodes every 10 minutes.
+ */
+final class PinUntilErrorChannel implements LimitedChannel {
```

Comment:
I wonder if we should make a specialized version of this for the case when there's exactly one upstream cos all the logic is pointless in that case (0 +1 % 1 is still 0)? Or am I over-optimizing?

Reply:
1 target is a very common case, I think it's worth optimizing for.
It's difficult to say _where_ we should do this optimization without implementing config reloading, if we continued with the 'recreate-everything' approach I'd prefer to simply bypass the load balancing channel entirely and wire the next LimitedChannel in the chain.